### ISSUE TITLE
Sketch (.ino) file monitoring isn't working for me in attach-and-watch

### ISSUE THREAD

**Issue by egnor** [robgonnella/ardi#55]
(Thanks for ardi! It is exactly what I was planning to write, but already written! 🙏)

As I understand it, `ardi attach-and-watch` is supposed to monitor the .ino file so if I edit and save, it will rebuild and reflash. However, this doesn't seem to work.

```
% ardi attach-and-watch -f arduino:avr:uno cap1188_tester.ino
ardi    Compiling...                                  fqbn="arduino:avr:uno" ske
tch=/home/egnor/source/machine/arduino/TESTS/cap1188_tester/cap1188_tester.ino
Sketch uses 5552 bytes (17%) of program storage space. Maximum is 32256 bytes.
Global variables use 532 bytes (25%) of dynamic memory, leaving 1516 bytes for l
ocal variables. Maximum is 2048 bytes.
ardi    Watching /home/egnor/source/machine/arduino/TESTS/cap1188
_tester/cap1188_tester.ino for changes 
ardi    Watching logs...                             
CAP1188 tester running
Product ID: 0x50
Manuf. ID: 0x5D
Revision: 0x83
CAP1188 init done
```

(The sketch continues running, and I still see output from it as it's generated.) Even though it says "Watching ... for changes", if I go save a change to cap1188_tester.ino in my editor, or even just `touch cap1188_tester.ino`, nothing happens. (Re-running the `ardi attach-and-watch` command does pick up the change, of course.)

My system is a relatively ordinary x86-64 desktop running Ubuntu 21.04. Further diagnostics available if it's helpful! I can also try `strace` and similar to try to see what's going on under the covers.

**Comment by robgonnella** [comment]
Thanks for reporting this! I will investigate this more and roll out a fix soon

**Comment by robgonnella** [comment]
I'm having trouble reproducing this although I did find a different bug with the timing of closing the serial connection before re-uploading. I'll be working on a patch for that next week. In the meantime do you think you could run the command with `-v` and post the logs?

```
ardi attach-and-watch -f arduino:avr:uno cap1188_tester.ino -v
```

**Comment by egnor** [comment]
(Will do when I get a chance!)

**Comment by egnor** [comment]
When I do, I see lots of build/upload related output, followed by:

```
ardi    Watching mpr121_tester.ino for changes       
ardi    Watching logs...                             
=== MPR121 tester starting ===
```

And then when I save the file in my editor (Vim), I see:

```
ardi   event: "mpr121_tester.ino": RENAME           
ardi   event: "mpr121_tester.ino": CHMOD            
ardi   event: "mpr121_tester.ino": REMOVE           
```

However, ardi doesn't reload the file. I guess you're just looking at WRITE events or something like that, versus move-and-replace operations? On subsequent writes I don't see further lines, presumably because it's watching the old file that has now been renamed and deleted? (I don't know the semantics of the file watching system you're using.)

**Comment by robgonnella** [comment]
Odd this is working fine for me using Vim and ubuntu (18.04, 20.04, & 21.04). I see `WRITE` events when saving. I'm not sure it makes sense to trigger recompilation and reuploading on `RENAME`, `CHMOD`, and `REMOVE` events.

**Comment by egnor** [comment]
The `fsnotify` docs are vague on semantics, but I _think_ you do have to look for `Create` and maybe `Rename` (but not `Chmod` or `Remove`) events in addition to `Write`.

I don't know why our versions of vim behave differently, but it's good editor save behavior to
1) write `foo.ino.tmp` with the new file
2) [optional] delete `foo.ino`
3) rename `foo.ino.tmp` to `foo.ino` (replacing the original if not already deleted)

That way the file gets updated atomically, so there's no chance of picking up a half-written file. However, when watching `foo.ino`, only `Create` and/or `Rename` (and maybe `Remove`) events will be generated.

On Linux, the underlying `inotify` API provides more granular events (https://man7.org/linux/man-pages/man7/inotify.7.html). Using that API, we'd want to watch `IN_CLOSE_WRITE` and `IN_MOVED_SELF` (but _not_ `IN_MODIFY`, to minimize the potential to read a half-written file). However, `fsnotify` (https://github.com/fsnotify/fsnotify/blob/master/inotify.go#L319) doesn't watch for `IN_CLOSE_WRITE` so we must watch `Write` and accept the race; fsnotify maps `IN_MOVED_SELF` to `Rename`, so we need to watch that. (Note, `IN_MOVED_TO` only applies when watching a directory.)

See https://github.com/fsnotify/fsnotify/issues/80 for additional discussion of the situation, and https://github.com/gohugoio/hugo/issues/5205 for a different project wrestling with this issue. "Best practices" for watching a single ordinary file aren't obvious!

**Comment by robgonnella** [comment]
Thanks for investigating this more and opening up an issue with `fsnotify`! After a bit of playing around with this locally here is what I've discovered on ubuntu

```
cp file.tmp file --> WRITE
mv file file.tmp --> RENAME
mv file.tmp file --> CHMOD, REMOVE
```

Based on this I've created a PR that watches for the `REMOVE` event as it seems as though that is the last event in an atomic update, and then checks if the target file still exists, if so it re-adds it to the watcher and continues with the update, if not we bail.

**Comment by robgonnella** [comment]
Updated in new beta version, you can try it out via `GO111MODULE=on go get github.com/robgonnella/ardi/v2@v2.3.0-beta.1`

**Comment by egnor** [comment]
Verified works for me! :tada:

**Comment by robgonnella** [comment]
Closing: Addressed in v2.3.0
